contracts/BN256G1.sol

```diff
@@ -0,0 +1,242 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity >=0.5.3 <0.7.0;
+
+import "elliptic-curve-solidity/contracts/EllipticCurve.sol";
+
+
+/**
+ * @title BN256G1 Curve Library
+ * @dev Library providing arithmetic operations over G1 in bn256.
+ * Provides additional methods like pairing and pairing_batch
+ * Heavily influenced by https://github.com/PhilippSchindler/ethdkg
+ * Calls to assembly are public and not external because assembly cannot be applied on calldata
+ * @author Witnet Foundation
+ */
+
+library BN256G1 {
+
+  // Generator coordinate `x` of the EC curve
+  uint256 public constant GX = 1;
+  // Generator coordinate `y` of the EC curve
+  uint256 public constant GY = 2;
+  // Constant `a` of EC equation
+  uint256 public constant AA = 0;
+  // Constant `b` of EC equation
+  uint256 public constant BB = 3;
+  // Prime number of the curve
+  uint256 public constant PP = 0x30644e72e131a029b85045b68181585d97816a916871ca8d3c208c16d87cfd47;
+  // Order of the curve
+  uint256 public constant NN = 0x30644e72e131a029b85045b68181585d2833e84879b9709143e1f593f0000001;
+
+  /// This is 0xf1f5883e65f820d099915c908786b9d3f58714d70a38f4c22ca2bc723a70f263, the last mulitple of the modulus before 2^256
+  uint256 public constant LAST_MULTIPLE_OF_PP_LOWER_THAN_2_256 = 0xf1f5883e65f820d099915c908786b9d3f58714d70a38f4c22ca2bc723a70f263;
+
+
+  /// @dev  computes P + Q
+  /// @param input: 4 values of 256 bit each
+  ///  *) x-coordinate of point P
+  ///  *) y-coordinate of point P
+  ///  *) x-coordinate of point Q
+  ///  *) y-coordinate of point Q
+  /// @return An array with x and y coordinates of P+Q.
+  function add(uint256[4] memory input) internal returns (uint256[2] memory) {
+    bool success;
+    uint256[2] memory result;
+    assembly {
+      // 0x06     id of precompiled bn256Add contract
+      // 0        number of ether to transfer
+      // 128      size of call parameters, i.e. 128 bytes total
+      // 64       size of call return value, i.e. 64 bytes / 512 bit for a BN256 curve point
+      success := call(not(0), 0x06, 0, input, 128, result, 64)
+    }
+    require(success, "bn256 addition failed");
+
+    return result;
+  }
+
+  /// @dev  computes computes P*k.
+  /// @param input: 3 values of 256 bit each:
+  ///  *) x-coordinate of point P
+  ///  *) y-coordinate of point P
+  ///  *) scalar k.
+  /// @return An array with x and y coordinates of P*k.
+  function multiply(uint256[3] memory input) internal returns (uint256[2] memory) {
```

Comment:
May it be better (gas wise) to return a tuple instead of an array?

Reply:
Might be, I just tried to make a common way for both libraries. In the BN256G2.sol we need to return 4 values, so we decided that it was worth returning them as a vector. This made us return a vector in every other method

Reply:
Some of BN256G2 functions were changed to apply these changes. However, some of them (internal usage though) have quite a few variables and stack gets too deep. In BN256G1 (as in this case), the precompiles expect an array and therefore it would not make sense to pass  tuples